

to hardErr()
/* Access undefined object to cause a hard error. */
    {
    string s;
    s[0] = 'H';
    }

to punt1()
    {
    punt("I'm not doing this for fun, but here goes.", "crashOut", 1);
    }

to punt0()
    {
    punt("I can't help it, I'm a failure.");
    }

array of var of to () punters = (punt1, punt0, hardErr);

to doSomethingFlaky(string when, var of to() punter)
    {
    for (i in 0 to 3)
        print("Testing " + i);
    print("oh rats...");
    punter();
    }

to flakarelli(int level)
    {
    int i = 0;
    print("welcome to flakarelli level " + level);
    for (v in punters)
	{
	try 
	   {
	   doSomethingFlaky("flaky test " + level, v);
	   }
	catch (m,s,level)
	   {
	   print("Caught punter " + i + ". I'm at level " + level);
	   }
	i += 1;
	}
    }

to doSomething()
    {
    string mess,src;
    int lev=2;
    try  // consider adding an optional time-out here.
	{
	doSomethingFlaky("now", hardErr);
	}
    catch (message,source,-1)
	{
        print("I guess it really was a flaky " + source + 
		". Here's the error message:");
	print(message);
	}
    print("**********************");
    print("And again with no catch");
    doSomethingFlaky("then", hardErr);
    }

flakarelli(-1);

try
   {
   print("Bravely running away.");
   punt("I just got scared");
   }
catch (m)
   {
   print("problem: " + m);
   }


